### ISSUE TITLE
[Feature]: LoRA support for Pixtral

### ISSUE THREAD

**Issue by spring-anth** [vllm-project/vllm#8802]
### 🚀 The feature, motivation and pitch

I have finetuned the linear layers of Pixtral on my own dataset and would like to host the LoRA adapters as is possible for Mistral. It would great if this would be supported in the future.

Related issue: #8685 as the base model I used for finetuning is the HF version

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by DarkLight1337** [comment]
LoRA support for VLMs in general is still WIP. cc @jeejeelee

**Comment by jeejeelee** [comment]
> LoRA support for VLMs in general is still WIP. cc @jeejeelee

Thanks for the ping, I should be able to complete the temporary solution for VL support of LoRA this week.

**Comment by jeejeelee** [comment]
@spring-anth  I have completed the integration of Pixtral support for LoRA, see: https://github.com/jeejeelee/vllm/tree/pixtral-support-lora. Could you please verify this locally? I don't have the enough resources to train the LoRA model myself.

**Comment by jeejeelee** [comment]
@spring-anth  Hi, which branch are you using? Is it `pixtral-support-lora`?

**Comment by spring-anth** [comment]
@jeejeelee You were right, I was on the wrong branch, silly mistake. Unfortunately I currently can't test for you if your change works as I trained pixtral with the transformers-compatible version. Therefore I can only use the LoRA weights for Pixtral once the transformers version of Pixtral is supported in vLLM (which is work in progress). My current workaround is merging the weights and transforming the model back to the vLLM compatible version.

**Comment by jeejeelee** [comment]
@spring-anth  Currently, vLLM only supports LoRA trained with PEFT

**Comment by spring-anth** [comment]
@jeejeelee Yes, that's not what I meant I did train with PEFT but the training is based on the HF Transformers version of Pixtral (https://huggingface.co/mistral-community/pixtral-12b) which uses a different structure than the vLLM supported version (https://huggingface.co/mistralai/Pixtral-12B-2409)

**Comment by tensimixt** [comment]
@jeejeelee does this work with https://github.com/vllm-project/vllm/pull/5036? If so can test this week if inference with mistralai/Pixtral-12b-2409 with Lora Adapter works. Or should i use another PR to test vllm inference of this? 

Thank you!

**Comment by jeejeelee** [comment]
> @jeejeelee does this work with #5036? If so can test this week if inference with mistralai/Pixtral-12b-2409 with Lora Adapter works. Or should i use another PR to test vllm inference of this?
> 
> Or I can just use this branch https://github.com/jeejeelee/vllm/tree/pixtral-support-lora
> 
> Will this work for python -m vllm.entrypoints.openai.api_server where model is set to pixtral and lora modules to pixtral lora adapter?
> 
> Thank you!

I thinki it should work, see: https://docs.vllm.ai/en/latest/models/lora.html#serving-lora-adapters

**Comment by tensimixt** [comment]
@jeejeelee when doing git checkout pixtral-support-lora  and then pip install -e . does it build correctly for you or it crashes?

**Comment by jeejeelee** [comment]
I also need to spend a long time, unless compiling on high-performance CPU servers.

**Comment by tensimixt** [comment]
@jeejeelee do you think pixtral-support-lora will work with this model https://huggingface.co/mistral-community/pixtral-12b where LoRA is loaded once model loaded with vllm serve. Or it must have LoRA trained on consolidated.safetensors from mistralai? 

Thank you!

**Comment by tensimixt** [comment]
> @jeejeelee You were right, I was on the wrong branch, silly mistake. Unfortunately I currently can't test for you if your change works as I trained pixtral with the transformers-compatible version. Therefore I can only use the LoRA weights for Pixtral once the transformers version of Pixtral is supported in vLLM (which is work in progress). My current workaround is merging the weights and transforming the model back to the vLLM compatible version.

Did you try to attempt loading mistralai-community pixtral with pixtral-support-lora branch and then load LoRA with vllm serve? 

Thank you!

**Comment by jeejeelee** [comment]
@tensimixt  Not yet. I will try it asps

**Comment by tensimixt** [comment]
> @tensimixt Not yet. I will try it asps

Thank you, this will help greatly!!

**Comment by jeejeelee** [comment]
> loading mistralai-community

@tensimixt  could you please provide the related link？

**Comment by tensimixt** [comment]
> > loading mistralai-community
> 
> @tensimixt could you please provide the related link？

It is this one https://huggingface.co/mistral-community/pixtral-12b 

Thank you!! You are amazing!

**Comment by jeejeelee** [comment]
@tensimixt   I have updated the pixtral-support-lora branch, and I can now run the following script successfully:
```python
vllm serve mistralai/Pixtral-12B-2409 --tokenizer-mode mistral --limit-mm-per-prompt 'image=4' --max-model-len 16384  --enable-lora
```
I am downloading this  model from `mistralai-community`

**Comment by jeejeelee** [comment]
I can't start  the vllm serve when using https://huggingface.co/mistral-community/pixtral-12b,  it raise the following error:
```shell
 `ERROR 11-12 06:47:47 engine.py:366] OSError: Found 0 files matching the pattern: re.compile('^tokenizer\\.model\\.v.*$|^tekken\\.json$'). Make sure that a Mistral tokenizer is present in ['model-00002-of-00006.safetensors', 'model-00005-of-00006.safetensors', 'preprocessor_config.json', 'model-00003-of-00006.safetensors', 'chat_template.json', 'tokenizer.json', 'generation_config.json', '.cache', 'model-00004-of-00006.safetensors', 'tokenizer_config.json', 'model.safetensors.index.json', '.gitattributes', 'config.json', 'model-00006-of-00006.safetensors', 'model-00001-of-00006.safetensors', 'README.md', 'special_tokens_map.json', 'processor_config.json'].`
```

**Comment by tensimixt** [comment]
> I can't start the vllm serve when using https://huggingface.co/mistral-community/pixtral-12b, it raise the following error:
> 
> ```shell
>  `ERROR 11-12 06:47:47 engine.py:366] OSError: Found 0 files matching the pattern: re.compile('^tokenizer\\.model\\.v.*$|^tekken\\.json$'). Make sure that a Mistral tokenizer is present in ['model-00002-of-00006.safetensors', 'model-00005-of-00006.safetensors', 'preprocessor_config.json', 'model-00003-of-00006.safetensors', 'chat_template.json', 'tokenizer.json', 'generation_config.json', '.cache', 'model-00004-of-00006.safetensors', 'tokenizer_config.json', 'model.safetensors.index.json', '.gitattributes', 'config.json', 'model-00006-of-00006.safetensors', 'model-00001-of-00006.safetensors', 'README.md', 'special_tokens_map.json', 'processor_config.json'].`
> ```

Thank you for updating the branch and trying to serve  https://huggingface.co/mistral-community/pixtral-12b using it. 
I tihnk it is trying to find tekken.json (but instead there is tokenizer.json in the repo) 

The current issue is that there is no way to create LoRA adapters for https://huggingface.co/mistralai/Pixtral-12B-2409
using transformers + peft approach (unless you know a way to make it work? ) 

And so the alternative is to create LoRA adapters using https://huggingface.co/mistral-community/pixtral-12b  which is compatible with transformers. 

Do you know if there is a way to create LoRA adapter for https://huggingface.co/mistralai/Pixtral-12B-2409 ?

Thank you!

**Comment by jeejeelee** [comment]
@tensimixt  I have no idea, maybe you can create a issue at https://github.com/mistralai/mistral-finetune
